l2geth/rollup/fees/rollup_fee.go

```diff
@@ -1,167 +1,278 @@
-// roundedL2GasLimit = ceilmod(l2GasLimit, 10_000)
-// l1Fee = l1GasPrice * l1GasLimit
-// l2Fee = l2GasPrice * roundedL2GasLimit
-// sum = l1Fee + l2Fee
-// scaled = sum / scalar
-// rounded = ceilmod(scaled, tenThousand)
-// roundedScaledL2GasLimit = roundedL2GasLimit / tenThousand
-// result = rounded + roundedScaledL2GasLimit
-// Note that for simplicity purposes, only the calldata is passed into this
-// function when in reality the RLP encoded transaction should be. The
-// additional cost is added to the overhead constant to prevent the need to RLP
-// encode transactions during calls to `eth_estimateGas`
-func EncodeTxGasLimit(data []byte, l1GasPrice, l2GasLimit, l2GasPrice *big.Int) *big.Int {
-	l1GasLimit := calculateL1GasLimit(data, overhead)
-	roundedL2GasLimit := Ceilmod(l2GasLimit, BigTenThousand)
-	l1Fee := new(big.Int).Mul(l1GasPrice, l1GasLimit)
-	l2Fee := new(big.Int).Mul(l2GasPrice, roundedL2GasLimit)
-	sum := new(big.Int).Add(l1Fee, l2Fee)
-	scaled := new(big.Int).Div(sum, bigFeeScalar)
-	rounded := Ceilmod(scaled, BigTenThousand)
-	roundedScaledL2GasLimit := new(big.Int).Div(roundedL2GasLimit, BigTenThousand)
-	result := new(big.Int).Add(rounded, roundedScaledL2GasLimit)
-	return result
-}
-
-func Ceilmod(a, b *big.Int) *big.Int {
-	remainder := new(big.Int).Mod(a, b)
-	if remainder.Cmp(common.Big0) == 0 {
-		return a
-	}
-	sum := new(big.Int).Add(a, b)
-	rounded := new(big.Int).Sub(sum, remainder)
-	return rounded
-}
-
-// DecodeL2GasLimit decodes the L2 gas limit from an encoded L2 gas limit
-func DecodeL2GasLimit(gasLimit *big.Int) *big.Int {
-	scaled := new(big.Int).Mod(gasLimit, BigTenThousand)
-	return new(big.Int).Mul(scaled, BigTenThousand)
-}
-
-func DecodeL2GasLimitU64(gasLimit uint64) uint64 {
-	scaled := gasLimit % tenThousand
-	return scaled * tenThousand
+// Message represents the interface of a message.
+// It should be a subset of the methods found on
+// types.Message
+type Message interface {
+	From() common.Address
+	To() *common.Address
+	GasPrice() *big.Int
+	Gas() uint64
+	Value() *big.Int
+	Nonce() uint64
+	Data() []byte
+}
+
+// StateDB represents the StateDB interface
+// required to compute the L1 fee
+type StateDB interface {
+	GetState(common.Address, common.Hash) common.Hash
+}
+
+// RollupOracle represents the interface of the in
+// memory cache of the gas price oracle
+type RollupOracle interface {
+	SuggestL1GasPrice(ctx context.Context) (*big.Int, error)
+	SuggestL2GasPrice(ctx context.Context) (*big.Int, error)
+	SuggestOverhead(ctx context.Context) (*big.Int, error)
+	SuggestScalar(ctx context.Context) (*big.Float, error)
+}
+
+// CalculateTotalFee will calculate the total fee given a transaction.
+// This function is used at the RPC layer to ensure that users
+// have enough ETH to cover their fee
+func CalculateTotalFee(tx *types.Transaction, gpo RollupOracle) (*big.Int, error) {
+	// Read the variables from the cache
```

Comment:
```
if tx.GasPrice().Cmp(common.Big0) == 0 {
    return new(big.Int), nil
}
```

Reply:
Noting a possible approach here, things work as is but this could abstract away some code